File: data-structures/non-linear/binary-search-tree/BinarySearchTree.cpp
#include <iostream>
#include <vector>
using namespace std;

class Node {
public:
  int data;
  Node *left;
  Node *right;

  Node(int value) {
    data = value;
    left = right = nullptr;
  }
};

Node *insert(Node *root, int value) {
  if (root == nullptr) {
    return new Node(value);
  }

  if (value < root->data) {
    root->left = insert(root->left, value);
  } else {
    root->right = insert(root->right, value);
  }

  return root;
}

Node *buildBST(vector<int> &vec) {
  Node *root = nullptr;

  for (int val : vec) {
    root = insert(root, val);
  }

  return root;
}

void inorderTraversal(Node *root) {
  if (root == nullptr) {
    return;
  }

  inorderTraversal(root->left);
  cout << root->data << " ";
  inorderTraversal(root->right);
}

bool search(Node *root, int key) {
  if (root == nullptr) {
    return false;
  }

  if (root->data == key) {
    return true;
  }

  if (root->data > key) {
    return search(root->left, key);
  } else {
    return search(root->right, key);
  }
}

Node *getInorderSuccessor(Node *root) {
  while (root != nullptr && root->left != nullptr) {
    root = root->left;
  }

  return root;
}

Node *deleteNode(Node *root, int key) {
  if (root == nullptr) {
    return nullptr;
  }

  if (root->data > key) {
    root->left = deleteNode(root->left, key);
  } else if (root->data < key) {
    root->right = deleteNode(root->right, key);
  } else {
    if (root->left == nullptr) {
      Node *temp = root->right;
      delete root;
      return temp;
    } else if (root->right == nullptr) {
      Node *temp = root->left;
      delete root;
      return temp;
    } else {
      Node *IS = getInorderSuccessor(root->right);
      root->data = IS->data;
      root->right = deleteNode(root->right, IS->data);
    }
  }

  return root;
}

File: data-structures/non-linear/binary-search-tree/BinarySearchTree.ts
import { TreeNode } from "../node/TreeNode";

function insert(root: TreeNode | null, value: number): TreeNode {
  if (root == null) {
    return new TreeNode(value);
  }

  if (value < root.getValue()) {
    root.setLeft(insert(root.getLeft(), value));
  } else {
    root.setRight(insert(root.getRight(), value));
  }

  return root;
}

function buildTree(preOrderArray: Array<number>): TreeNode | null {
  let node: TreeNode | null = null;

  for (const value of preOrderArray) {
    node = insert(node, value);
  }

  return node;
}

function preOrderTraversal(root: TreeNode | null): void {
  if (root == null) {
    return;
  }

  console.log(`${root.getValue()}`);
  preOrderTraversal(root.getLeft());
  preOrderTraversal(root.getRight());
}

function inOrderTraversal(root: TreeNode | null): void {
  if (root == null) {
    return;
  }

  inOrderTraversal(root.getLeft());
  console.log(`${root.getValue()}`);
  inOrderTraversal(root.getRight());
}

function postOrderTraversal(root: TreeNode | null): void {
  if (root == null) {
    return;
  }

  postOrderTraversal(root.getLeft());
  postOrderTraversal(root.getRight());
  console.log(`${root.getValue()}`);
}

function search(root: TreeNode | null, value: number): boolean {
  if (root == null) {
    return false;
  }

  if (root.getValue() == value) {
    return true;
  }

  if (root.getValue() < value) {
    return search(root.getLeft(), value);
  } else {
    return search(root.getRight(), value);
  }
}

function inOrderSuccessor(root: TreeNode | null): TreeNode | null {
  while (root != null && root.getLeft() != null) {
    root.setLeft(root.getLeft());
  }

  return root;
}

function deleteNode(root: TreeNode | null, value: number): TreeNode | null {
  if (root == null) {
    return null;
  }

  if (root.getValue() > value) {
    root.setLeft(deleteNode(root.getLeft(), value));
  } else if (root.getValue() < value) {
    root.setRight(deleteNode(root.getRight(), value));
  } else {
    if (root.getLeft() == null) {
      return root.getRight();
    } else if (root.getRight() == null) {
      return root.getLeft();
    }

    const successor: TreeNode | null = inOrderSuccessor(root.getRight());
    //@ts-expect-error There is no possible way to get null in inOrderSuccessor
    root.setValue(successor.getValue());
    //@ts-expect-error There is no possible way to get null in inOrderSuccessor
    root.setRight(deleteNode(root.getRight(), successor?.getValue()));
  }

  return root;
}
